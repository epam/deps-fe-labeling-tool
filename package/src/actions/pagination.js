import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
