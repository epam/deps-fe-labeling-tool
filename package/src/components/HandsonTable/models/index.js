
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
