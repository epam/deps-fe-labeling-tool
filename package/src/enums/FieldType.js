const FieldType = {
  CHECKMARK: 'checkmark',
  ENUM: 'enum',
  LIST: 'list',
  PAIR: 'pair',
  STRING: 'string',
  TABLE: 'table',
  DATE: 'date'
}

const RESOURCE_FIELD_TYPE = {
  [FieldType.STRING]: 'String',
  [FieldType.ENUM]: 'Enumeration',
  [FieldType.CHECKMARK]: 'Checkbox',
  [FieldType.LIST]: 'List',
  [FieldType.PAIR]: 'Key Value Pair',
  [FieldType.TABLE]: 'Table',
  [FieldType.DATE]: 'Date'
}

export {
  FieldType,
  RESOURCE_FIELD_TYPE
}
