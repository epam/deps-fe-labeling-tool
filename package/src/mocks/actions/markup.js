const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
