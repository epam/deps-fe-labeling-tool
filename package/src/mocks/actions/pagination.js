const openPage = jest.fn((page) => ({
  type: 'MOCK_OPEN_PAGE_ACTION',
  payload: page
}))

const mockPaginationActions = {
  openPage
}

export {
  mockPaginationActions
}
