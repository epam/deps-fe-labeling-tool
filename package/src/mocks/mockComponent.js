const mockComponent = (componentName, namedExport = true, fn = () => { }) => {
  Object.defineProperty(fn, 'name', { value: componentName })
  return !namedExport ? fn : { [componentName]: fn }
}

export {
  mockComponent
}
