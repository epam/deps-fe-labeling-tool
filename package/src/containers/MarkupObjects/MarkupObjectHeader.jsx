import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { clearSelection } from '@/actions/markup'
import { openPage } from '@/actions/pagination'
import { changeTool } from '@/actions/tools'
import { setActiveSidebar } from '@/actions/ui'
import { TextHighlighter } from '@/components/TextHighlighter'
import { SidebarContent } from '@/enums/SidebarContent'
import { Tool } from '@/enums/Tool'
import { labelShape } from '@/models/Label'
import { tableShape } from '@/models/Table'
import { pageSelectedLabelsSelector, pageSelectedTablesSelector } from '@/selectors/markup'
import { currentPageSelector } from '@/selectors/pagination'
import { selectedToolSelector } from '@/selectors/tools'
import { markupObjectsFilterSelector } from '@/selectors/ui'
import { childrenShape } from '@/utils/propTypes'
import { replaceAll } from '@/utils/string'
import { TextWrapper, TitleWithIconWrapper, PageSpan, NameSpan } from './MarkupObjectHeader.styles'

const RESOURCE_PAGE_1 = 'page {0}'

const MarkupObjectHeader = ({
  badge,
  collapseIcon,
  icon,
  name,
  pagesTitle,
  unassigned,
  page,
  markupObjectFilter,
  onHeaderClick,
  currentPage,
  clearSelection,
  openPage,
  changeTool,
  selectedTool,
  selectedLabels,
  selectedTables,
  setActiveSidebar
}) => {
  const onClick = () => {
    if (selectedTool !== Tool.POINTER) {
      changeTool(Tool.POINTER)
    }

    if (selectedLabels.length || selectedTables.length) {
      clearSelection(currentPage)
    }

    setActiveSidebar(SidebarContent.MARKUP)

    if (currentPage !== page && page !== undefined) {
      openPage(page)
    }

    onHeaderClick?.()
  }

  return (
    <TextWrapper
      $unassigned={unassigned}
      onClick={onClick}
    >
      {badge}
      {collapseIcon}
      <TitleWithIconWrapper>
        {icon}
        <NameSpan>
          <TextHighlighter
            searchTerm={markupObjectFilter}
            text={name}
          />
        </NameSpan>
      </TitleWithIconWrapper>
      {pagesTitle && <PageSpan>{pagesTitle}</PageSpan>}
      {page && <PageSpan>{replaceAll(RESOURCE_PAGE_1, page)}</PageSpan>}
    </TextWrapper>
  )
}

MarkupObjectHeader.propTypes = {
  badge: childrenShape,
  collapseIcon: childrenShape,
  icon: childrenShape,
  pagesTitle: PropTypes.string,
  setActiveSidebar: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  page: PropTypes.number,
  unassigned: PropTypes.bool,
  markupObjectFilter: PropTypes.string,
  onHeaderClick: PropTypes.func,
  currentPage: PropTypes.number.isRequired,
  openPage: PropTypes.func.isRequired,
  clearSelection: PropTypes.func.isRequired,
  changeTool: PropTypes.func.isRequired,
  selectedTool: PropTypes.oneOf(
    Object.values(Tool)
  ).isRequired,
  selectedLabels: PropTypes.arrayOf(labelShape).isRequired,
  selectedTables: PropTypes.arrayOf(tableShape).isRequired
}

const mapStateToProps = (state) => ({
  markupObjectFilter: markupObjectsFilterSelector(state),
  currentPage: currentPageSelector(state),
  selectedTool: selectedToolSelector(state),
  selectedLabels: pageSelectedLabelsSelector(state),
  selectedTables: pageSelectedTablesSelector(state)
})

const mapDispatchToProps = {
  openPage,
  clearSelection,
  changeTool,
  setActiveSidebar
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(MarkupObjectHeader)

export {
  ConnectedComponent as MarkupObjectHeader
}
