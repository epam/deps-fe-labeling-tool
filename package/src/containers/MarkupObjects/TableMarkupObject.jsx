import React, { useCallback } from 'react'
import { PropTypes } from 'prop-types'
import { connect } from 'react-redux'
import { selectTables } from '@/actions/markup'
import { fieldShape } from '@/models/Field'
import { Markup, markupShape } from '@/models/Markup'
import { tableShape } from '@/models/Table'
import { markupSelector } from '@/selectors/markup'
import { MarkupObjectHeader } from './MarkupObjectHeader'
import { MARKUP_OBJECT_TYPE_TO_ICON } from './MarkupObjectToIcon'

const TableMarkupObject = ({
  markup,
  table,
  field,
  selectTables
}) => {
  const objectPage = Markup.getPageByObjectUid(markup, table.uid)

  const onHeaderClick = useCallback(() => {
    selectTables(objectPage, [table])
  }, [objectPage, selectTables, table])

  return (
    <MarkupObjectHeader
      key={table.uid}
      page={objectPage}
      unassigned={table.fieldCode === ''}
      name={`${field.name} ${table?.index ?? ''}`}
      onHeaderClick={onHeaderClick}
      icon={MARKUP_OBJECT_TYPE_TO_ICON[table.typeName]}
    />
  )
}

TableMarkupObject.propTypes = {
  markup: markupShape,
  table: tableShape,
  field: fieldShape,
  selectTables: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  markup: markupSelector(state)
})

const mapDispatchToProps = {
  selectTables
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(TableMarkupObject)

export {
  ConnectedComponent as TableMarkupObject
}
