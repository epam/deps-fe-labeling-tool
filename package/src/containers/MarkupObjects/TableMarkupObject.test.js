import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockUuid } from '@/mocks/mockUuid'
import { mockMarkupSelectors } from '@/mocks/selectors/markup'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import React from 'react'
import { shallow } from 'enzyme'
import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { MarkupObjectHeader } from './MarkupObjectHeader'
import { TableMarkupObject } from './TableMarkupObject'

jest.mock('uuid', () => mockUuid)

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/markup', () => mockMarkupSelectors)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('./MarkupObjectHeader', () => mockComponent('MarkupObjectHeader'))

const mockObjectPage = 1

const {
  WrappedComponent,
  mapStateToProps
} = TableMarkupObject

describe('Component: TableMarkupObject', () => {
  describe('mapStateToProps', () => {
    it('should pass state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      const expectedProps = {
        markup: mockMarkupSelectors.markupSelector()
      }

      expect(props).toEqual(expectedProps)
    })
  })

  describe('component', () => {
    let defaultProps, wrapper

    beforeEach(() => {
      defaultProps = {
        field: new Field('MOCK_FIELD_4', 'test Field Name', FieldType.TABLE),
        table: mockMarkupSelectors.markupSelector()[mockObjectPage].tables[0],
        markup: mockMarkupSelectors.markupSelector(),
        selectTables: jest.fn()
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render correct layout', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call selectTables props with correct values when table is selected', () => {
      const objectHeaderProps = wrapper.find(MarkupObjectHeader).props()
      objectHeaderProps.onHeaderClick()
      expect(defaultProps.selectTables).nthCalledWith(1, mockObjectPage, [defaultProps.table])
    })
  })
})
