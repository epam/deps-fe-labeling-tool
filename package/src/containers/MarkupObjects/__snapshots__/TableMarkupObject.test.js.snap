// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TableMarkupObject component should render correct layout 1`] = `
<MarkupObjectHeader
  icon={
    <IconSize23>
      <TableIcon />
    </IconSize23>
  }
  key="5"
  name="test Field Name "
  onHeaderClick={[Function]}
  page={1}
  unassigned={false}
/>
`;
